### PR TITLE
Add rendering tests

### LINKED ISSUE #2923: Verify correctness of rendered map in tests (image comparison)
Right now, there are no tests that verify if the rendered map is correct. It would be good to have unit tests for the rendering, and also tests that check if the examples stay the same.

Related links:

* [Wraith — A responsive screenshot comparison tool](https://github.com/BBC-News/wraith)
* [Resemble.js](https://github.com/Huddle/Resemble.js)
* [js-imagediff](https://github.com/HumbleSoftware/js-imagediff)
* [mapbox-gl-test-suite](https://github.com/mapbox/mapbox-gl-test-suite)
* Three.js: https://github.com/mrdoob/three.js/issues/3871 https://github.com/mrdoob/three.js/pull/3873 https://github.com/danielribeiro/threejs-test https://github.com/mrdoob/three.js/pull/3976
* SlimerJS: https://github.com/mnem/travisci-webgl-test https://github.com/GoodBoyDigital/pixi.js/issues/216
* http://stackoverflow.com/questions/14015073/server-side-fallback-rendering

### PR DESCRIPTION
**Pull request by tsauerwein**
Currently large parts of our rendering code lack tests. This PR aims to close these gaps by providing an infrastructure for rendering tests. The idea is that the canvas of a rendered map is compared with a reference image using [Resemble.js](http://huddle.github.io/Resemble.js/).

For being able to run the tests for the canvas and WebGL renderer on Travis, [SlimerJS](http://slimerjs.org/) is used instead of [PhantomJS](http://phantomjs.org/), that is used for the current unit tests.

Like with the current unit tests, the rendering tests can be run in a browser or from the command-line (`./build.py test-rendering`). If a test fails, that is a mismatch factor between the reference image and the rendered map exceeds a certain threshold, an image showing the difference can be displayed.

rendering-tests

Still to do:

- [x] Write more tests (for different layer types).
- [ ] Make sure no window is opened when running the tests from the command line.

Closes #2923

### CODE REVIEW

**Comment by tsauerwein** [comment]
This PR would now be ready to review. For sure, the new tests do no cover everything of our rendering logic, but it's a start.

There is one open issue that I wasn't able to resolve: When running the tests locally with `./build.py test-rendering`, SlimerJS opens a window. It [should](http://docs.slimerjs.org/current/installation.html#having-a-headless-slimerjs) be possible to run the tests headless with xvfb (like it is done on Travis). But I wasn't able to enable WebGL for xvfb on my machine. So for now you'll see the SlimerJS window when running the tests. I don't know if others consider this as a blocker?

**Comment by elemoine** [comment]
Looks very good to me.

**Comment by bartvde** [comment]
I don't consider it a blocker

Thanks for this great achievement.

**Comment by elemoine** [comment]
> I don't consider it a blocker

Yeah, me neither.

**Comment by marcjansen** [comment]
With me it's three :smile: I am very happy once this is in.

:+1: for merging

File: test_rendering/spec/ol/layer/image.test.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        console.log('start ' + imagesLoading + ' ' + imagesLoaded);
+        imagesLoading++;
+      });
+      source.on('imageloadend', function(event) {
+        console.log('end ' + imagesLoading + ' ' + imagesLoaded);
+        imagesLoaded++;
+        update();
+      });
+      source.on('imageloaderror', function(event) {
+        console.log('error ' + imagesLoading + ' ' + imagesLoaded);
```
Comment (tschaub) [inline test_rendering/spec/ol/layer/image.test.js]: Would it make sense to call `onImagesLoaded` with an `Error` here?  Then the tests could call `done` with the same.  Then the `console.log` calls could probably be removed.

**Comment by tschaub** [comment]
This is so cool.  Thanks for all the great work on this @tsauerwein.  One minor question/comment above, but I also think this looks good to merge.

File: test_rendering/spec/ol/layer/image.test.js (hunk 1 above)
Comment (tsauerwein) [inline test_rendering/spec/ol/layer/image.test.js]: Good point! The tests should fail if the image can not be loaded.

**Comment by tsauerwein** [comment]
Thanks for your feedback, merging.

**Comment by tschaub** [comment]
> coverage/coveralls — Coverage increased (+4.23%)

Gold star for @tsauerwein!

**Comment by bartvde** [comment]
wow we are now at 72% should we increase our threshold to 70% already ;-) ?

**Comment by elemoine** [comment]
I wonder why adding rendering tests increased the code coverage. Can someone please explain?

**Comment by bartvde** [comment]
Looking at https://coveralls.io/builds/2272184 I also can't find a very logical explanation @marcjansen any ideas?

**Comment by tsauerwein** [comment]
The +4.23% do not come from the rendering tests. They are not included in the coverage.

I made a change to [tasks/server.js](https://github.com/openlayers/ol3/commit/8e0c21eb580bbd7a659fa5088b9c145030f9b872#diff-36350e8f94e2c54444cb93e007e1a6a0), maybe it's related to that.

**Comment by elemoine** [comment]
> I made a change to tasks/server.js, maybe it's related to that.

Which would be strange, right?

**Comment by bartvde** [comment]
Ah so maybe we should limit coverage to the src directory?

**Comment by elemoine** [comment]
> Ah so maybe we should limit coverage to the src directory?

Would make sense to me.

**Comment by marcjansen** [comment]
Strange indeed. It may be that the increase is due to the previous commit in that pr (which failed the testsuite)... I am not sure though

Coverage is only relevant to files in the src directory.

**Comment by marcjansen** [comment]
It may also be that the new lib entries for the closure manager already raise the bar. Also it may be that we reach different code paths now that we run xvfb and have DISPLAY on travis. I'll see if I find time to understand more...

**Comment by marcjansen** [comment]
The coveralls report for this pr says: 14072 OF 19484 RELEVANT LINES COVERED, and 215 files covered.

The [previous one](https://coveralls.io/builds/2272123) has: 14706 OF 21630 RELEVANT LINES COVERED, and 248 files covered.

So we lost some covered files/lines, which raises the percentage.

Not sure wat caused this.

**Comment by tsauerwein** [comment]
I was just realizing the same. One file that disappeared is ` /src/ol/control/fullscreencontrol.js`:
https://coveralls.io/builds/2272123/source?filename=src%2Fol%2Fcontrol%2Ffullscreencontrol.js#L29

And it seems like there is no test for this file.

**Comment by marcjansen** [comment]
I am on my mobile, and cannot dig deeper, sorry...

**Comment by tsauerwein** [comment]
Before, the loader server returned all src files. Now it only returns the src files that are actually referenced by a test. And it seems like Istanbul does not include the files, that were not loaded during the test run. Though they should be instrumented?!

**Comment by marcjansen** [comment]
That's it, nice @tsauerwein!

Every source is instrumented, but when an instrumented file isn't loaded when testing, istanbul does not include a coverage point in the data it sends to coveralls.

We might want to unconditionally load every source prior toactually testing, so that the baseline is correct.

**Comment by marcjansen** [comment]
Even better would be if we had testfiles for every source, even if they weren't actually testing sth. for now.

It should be doable though to have tests for 'can be constructed' or sth.

**Comment by marcjansen** [comment]
@tsauerwein: Awarding you the silver medal in debugging coverages, in addition to zhe gold thingy @tschaub gave you

**Pull request merged**